### PR TITLE
Feature/6240 signed key2

### PR DESCRIPTION
**Pull request by meskio**
* Some fixes
* upgrade keys if not successfully used
* upgrade key if signed by old key
* port to enum34

### CODE REVIEW

File: src/leap/keymanager/openpgp.py [hunk 1]
```diff
@@ -447,23 +434,30 @@ def _put_key_doc(self, key):
...
+                        gpg.import_keys(oldkey.key_data)
                         gpg.import_keys(key.key_data)
                         gpgkey = gpg.list_keys(secret=key.private).pop()
-                        key = _build_key_from_gpg(
+                        mergedkey = self._build_key_from_gpg(
                             gpgkey,
                             gpg.export_keys(gpgkey['fingerprint'],
                                             secret=key.private))
-                    doc.set_json(key.get_json())
+                    mergedkey.validation = max(
```
Comment (andrejb) [inline src/leap/keymanager/openpgp.py]: If the new key has smaller validation, why should the merged key be as valid as the old key?

File: src/leap/keymanager/openpgp.py [hunk 2]
```diff
@@ -447,23 +434,30 @@ def _put_key_doc(self, key):
...
+                        [key.validation, oldkey.validation])
+                    mergedkey.last_audited_at = oldkey.last_audited_at
+                    mergedkey.refreshed_at = key.refreshed_at
+                    mergedkey.encr_used = key.encr_used or oldkey.encr_used
```
Comment (andrejb) [inline src/leap/keymanager/openpgp.py]: If the new key was never used for encryption or signed, why should it have the same status as the new key?

Comment (andrejb) [inline src/leap/keymanager/openpgp.py]: Maybe we should a reference here to the validation specification?

File: pkg/requirements.pip [hunk 3]
```diff
@@ -1,7 +1,8 @@
 leap.common>=0.3.7
+leap.soledad.client>=0.6
```
Comment (andrejb) [inline pkg/requirements.pip]: Why 0.6 ?

File: src/leap/keymanager/validation.py [hunk 4]
```diff
@@ -24,17 +24,17 @@
...
-    "Historically_Auditing",
-    "Known_Key",
+ValidationLevel = IntEnum("ValidationLevel",
+    "Weak_Chain "
+    "Provider_Trust "
+    "Provider_Endorsement "
+    "Third_Party_Endorsement "
+    "Third_Party_Consensus "
+    "Historically_Auditing "
+    "Known_Key "
```
Comment (andrejb) [inline src/leap/keymanager/validation.py]: Why the spaces in the end of these strings?

File: src/leap/keymanager/openpgp.py (hunk 1 above)
Comment (meskio) [inline src/leap/keymanager/openpgp.py]: If an attacker manages to push you the same key, but with lower validation level that should not allow her to downgrade your validation level for this key.

One example:
1. Alice checks the fingerprint of Bob, and puts the key in her keymanager with validation level _fingerprint_.
2. Eve sends an email to Alice impersonating Bob with the Bob public key as attachment.
3. Alice gets Bob's key from the fake email of Eve and downgrade the validation level to _weak chain_.
4. Bob's key expires and Eve impersonates Bob again sending an email to Alice with a new public key that she has created, before Bob has time to update his expiration date.
5. Alice accepts this new public key as it has the same validation level and now communicates with Bob encrypting everything with Eve's key.

File: src/leap/keymanager/openpgp.py (hunk 2 above)
Comment (meskio) [inline src/leap/keymanager/openpgp.py]: I don't understand the first question. This code keeps the used flag as True if the old or new key has it True, if both are False it set's it up as False. I use that to push a key with the flag set from leap.mail.

You are right that I can link the validation specification, but is nowhere besides our mailing list yet.

File: pkg/requirements.pip (hunk 3 above)
Comment (meskio) [inline pkg/requirements.pip]: @kalikaneko told me to keep it with 0.6 and add 0.7 to the *VERSION_COMPAT* so setup.py install keeps working. I think the normal procedure is to bump the requirements when the new release comes out.

File: src/leap/keymanager/validation.py (hunk 4 above)
Comment (meskio) [inline src/leap/keymanager/validation.py]: I migrated it to enum34, where the values of the enum are not separated with _,_ as in enum but they are one string with the values separated with spaces. I think is easier to read it if I keep the string like that changing lines for each value, but it's not pep8 compliant.

File: pkg/requirements.pip (hunk 3 above)
Comment (kalikaneko) [inline pkg/requirements.pip]: yes I did :P

**Comment by kalikaneko** [comment]
it might be good to publish the spec to a repo, don't you think? so we can document the code.
@elijh @meskio

**Comment by kalikaneko** [comment]
now that I realize, I'd vetto the inclusion of soledad-client as a dep.
That means that mx will pull it

**Pull request merged**